src/tools/rbd_mirror/image_map/Policy.cc

```diff
@@ -0,0 +1,239 @@
+  dout(20) << ": global_image_id=" << global_image_id << dendl;
+
+  RWLock::WLocker map_lock(m_map_lock);
+
+  std::string instance_id = _lookup(global_image_id);
+  if (instance_id == UNMAPPED_INSTANCE_ID) {
+    return false;
+  }
+
+  m_map[instance_id].erase(global_image_id);
+  m_shuffled_timestamp.erase(global_image_id);
+  return true;
+}
+
+template <typename I>
+void Policy<I>::remap(const std::string &global_image_id,
+                      const std::string &source_instance_id,
+                      const std::string &target_instance_id, utime_t remap_time) {
+  dout(20) << ": global_image_id=" << global_image_id << ", source_instance_id="
+           << source_instance_id << ", target_instance_id=" << target_instance_id
+           << dendl;
+
+  assert(m_map_lock.is_wlocked());
+
+  m_map[target_instance_id].emplace(global_image_id);
+  m_map[source_instance_id].erase(global_image_id);
+  m_shuffled_timestamp[global_image_id] = remap_time;
+}
+
+template <typename I>
+bool Policy<I>::instance_present(const std::string &instance_id) {
+  dout(20) << ": instance_id=" << instance_id << dendl;
+
+  RWLock::RLocker map_lock(m_map_lock);
+  return m_map.find(instance_id) != m_map.end();
+}
+
+template <typename I>
+void Policy<I>::add_instances(const std::vector<std::string> &instances) {
+  dout(20) << dendl;
+
+  assert(m_map_lock.is_wlocked());
+
+  for (auto const &instance : instances) {
+    dout(10) << ": adding instance_id=" << instance << dendl;
+    m_map.emplace(instance, std::set<std::string>{});
+  }
+}
+
+template <typename I>
+void Policy<I>::remove_instances(const std::vector<std::string> &instances) {
+  dout(20) << dendl;
+
+  assert(m_map_lock.is_wlocked());
+
+  for (auto const &instance : instances) {
+    dout(10) << ": removing instance_id=" << instance << dendl;
+    if (!m_map[instance].empty()) {
+      dout(5) << ": instance_id=" << instance << " still has " << m_map[instance].size()
+              << " mapped image(s), continuing removal" << dendl;
+    }
+
+    m_map.erase(instance);
+  }
+}
+
+template <typename I>
+void Policy<I>::shuffle_add_instances(const std::vector<std::string> &instances,
+                                      std::map<std::string, std::pair<std::string, std::string> > *remap_ids) {
+  dout(20) << ": adding " << instances.size() << " instance(s)" << dendl;
+
+  assert(m_map_lock.is_wlocked());
+
+  add_instances(instances);
+  do_shuffle_add_instances(instances, remap_ids);
+
+  utime_t now = ceph_clock_now();
+  CephContext *cct = reinterpret_cast<CephContext *>(m_ioctx.cct());
+  int migration_throttle = cct->_conf->get_val<int64_t>("rbd_mirror_image_policy_migration_throttle");
+
+  dout(10) << ": migration_throttle=" << migration_throttle << " sec(s)" << dendl;
+
+  for (auto it = (*remap_ids).begin(); it != (*remap_ids).end();) {
+    const std::string &global_image_id = (*it).first;
+
+    if (!can_shuffle_image(now, m_shuffled_timestamp[global_image_id], migration_throttle)) {
```

Comment:
If the image cannot be shuffled -- the ```SimplyPolicy``` should have excluded it from its decision making

Reply:
Shouldn't this override (based to last shuffled time) be decided by the policy class rather than its implementers? Here, `SimplePolicy` just returns a set of image ids based on its balancing algorithm.

Reply:
Yeah -- but if this class is going to override the decision, then the policy will be sub-optimal since it might be picking an image that cannot move to move when instead it could pick an image that can move instead.